activerecord/lib/active_record/testing/query_assertions.rb

```diff
@@ -5,29 +5,76 @@ module Assertions
     module QueryAssertions
       # Asserts that the number of SQL queries executed in the given block matches the expected count.
       #
+      #   # Check for exact number of queries
       #   assert_queries(1) { Post.first }
       #
+      #   # Check for any number of queries
+      #   assert_queries { Post.first }
+      #
       # If the +:matcher+ option is provided, only queries that match the matcher are counted.
       #
       #   assert_queries(1, matcher: /LIMIT \?/) { Post.first }
       #
-      def assert_queries(expected_count, matcher: nil, &block)
+      # If the +:ignore_none+ option is provided, any queries (including transaction and schema related)
+      #   that match the matcher are counted.
+      #
+      #   assert_queries(ignore_none: true) { Post.columns }
+      #
+      def assert_queries(expected_count = nil, matcher: nil, ignore_none: false, &block)
         ActiveRecord::Base.connection.materialize_transactions
 
-        queries = []
-        callback = lambda do |*, payload|
-          queries << payload[:sql] if %w[ SCHEMA TRANSACTION ].exclude?(payload[:name]) && (matcher.nil? || payload[:sql].match(matcher))
-        end
-        ActiveSupport::Notifications.subscribed(callback, "sql.active_record") do
+        counter = SQLCounter.new
+        ActiveSupport::Notifications.subscribed(counter, "sql.active_record") do
           result = _assert_nothing_raised_or_warn("assert_queries", &block)
-          assert_equal expected_count, queries.size, "#{queries.size} instead of #{expected_count} queries were executed. Queries: #{queries.join("\n\n")}"
+          queries = ignore_none ? counter.log_all : counter.log
+
+          if expected_count.nil? && matcher
+            failed_patterns = []
+            matcher = Array(matcher)
+            matcher.each do |pattern|
+              failed_patterns << pattern unless queries.any? { |query| pattern === query }
+            end
+            assert_predicate failed_patterns, :empty?, "Query pattern(s) #{failed_patterns.map(&:inspect).join(', ')} not found.#{queries.empty? ? '' : "\nQueries:\n#{queries.join("\n")}"}"
+          else
+            matched_queries =
+              if matcher
+                matcher = Array(matcher)
+                queries.select { |query| matcher.any? { |pattern| pattern === query } }
+              else
+                queries
+              end
+
+            if expected_count.nil?
+              assert_operator matched_queries.size, :>=, 1, "1 or more queries expected, but none were executed."
+            else
+              assert_equal expected_count, matched_queries.size, "#{matched_queries.size} instead of #{expected_count} queries were executed. Queries: #{matched_queries.join("\n\n")}"
+            end
+          end
```

Comment:
Would it be possible to support `Range` instances too?

Something like `assert_queries(2..3) { ... }`. It'd also support endless Ranges: `assert_queries(..3) { ... }`.

```suggestion
            case expected_count
            when NilClass
              assert_operator matched_queries.size, :>=, 1, "1 or more queries expected, but none were executed."
            when Range
              assert_includes expected_count, matched_queries.size, "expected #{matched_queries.size} executed queries to be within #{expected_count}. Queries: #{matched_queries.join("\n\n")}"
            else
              assert_equal expected_count, matched_queries.size, "#{matched_queries.size} instead of #{expected_count} queries were executed. Queries: #{matched_queries.join("\n\n")}"
            end
          end
```

Reply:
Can you provide a use case for this?

Reply:
The main motivation is flexibility and maintenance.

I'm imagining introducing this assertion into an existing application test suite in the hopes that it'd drive efforts to minimize queries (either as an optimization or as a reaction to issues like N+1s). In this hypothetical application, I might introduce the assertion with a specific query, pass CI, then merge my change. A teammate might make a change that would make a query and the test might fail. They could update the assertion to use the more correct number, pass CI, then merge their change.

Depending on the goals of the team using this assertion that type of churn might start a virtuous cycle that forces the team to change the test's number for necessary queries. Alternatively, it might change often enough to develop a negative perception, feel onerous, and ultimately get removed.

In cases where the assertion is used as an optimization, teams might be interested in providing a `Range` (like "fewer than 10" or `..10`) that they could use as a arbitrary goal.

Personally, my preference is for the specific, exact count version. In my experiences, constant churn can deter that style of usage.

Reply:
> Personally, my preference is for the specific, exact count version.

This is my preference too. Let's see what other people say.